google/cloud/bigquery/_job_helpers.py

```diff
@@ -0,0 +1,173 @@
+
+
+def query_jobs_insert(
+    client: Client,
+    query: str,
+    job_config: job.QueryJobConfig,
+    job_id: str,
+    job_id_prefix: str,
+    location: str,
+    project: str,
+    retry: retries.Retry,
+    timeout: float,
+    job_retry: retries.Retry,
+):
+    job_id_given = job_id is not None
+    job_id_save = job_id
+    job_config_save = job_config
+
+    def do_query():
+        # Make a copy now, so that original doesn't get changed by the process
+        # below and to facilitate retry
+        job_config = copy.deepcopy(job_config_save)
+
+        job_id = make_job_id(job_id_save, job_id_prefix)
+        job_ref = job._JobReference(job_id, project=project, location=location)
+        query_job = job.QueryJob(job_ref, query, client=client, job_config=job_config)
+
+        try:
+            query_job._begin(retry=retry, timeout=timeout)
+        except core_exceptions.Conflict as create_exc:
+            # The thought is if someone is providing their own job IDs and they get
+            # their job ID generation wrong, this could end up returning results for
+            # the wrong query. We thus only try to recover if job ID was not given.
+            if job_id_given:
+                raise create_exc
+
+            try:
+                query_job = client.get_job(
+                    job_id,
+                    project=project,
+                    location=location,
+                    retry=retry,
+                    timeout=timeout,
+                )
+            except core_exceptions.GoogleAPIError:  # (includes RetryError)
+                raise create_exc
+            else:
+                return query_job
+        else:
+            return query_job
+
+    future = do_query()
+    # The future might be in a failed state now, but if it's
+    # unrecoverable, we'll find out when we ask for it's result, at which
+    # point, we may retry.
+    if not job_id_given:
+        future._retry_do_query = do_query  # in case we have to retry later
+        future._job_retry = job_retry
+
+    return future
+
+
+def query_jobs_query(
+    client: Client,
+    query: str,
+    job_config: job.QueryJobConfig,
+    location: str,
+    project: str,
+    retry: retries.Retry,
+    timeout: float,
+    job_retry: retries.Retry,
+):
+    # TODO: Validate that destination is not set.
+
+    request_body = {}
+    job_config_resource = job_config.to_api_repr()
+
+    # Transform from Job resource to QueryRequest resource.
+    # Most of the keys in job.configuration.query are in common
+    request_body.update(job_config_resource["configuration"]["query"])
+    request_body["location"] = location
+    request_body["labels"] = job_config.labels
+    request_body["dryRun"] = job_config.dry_run
+
+    # Subtract a buffer for context switching, network latency, etc.
+    request_body["timeoutMs"] = max(0, int(1000 * (timeout - _TIMEOUT_BUFFER_SECS)))
+
+    def do_query():
+        request_body["requestId"] = make_job_id(None)
```

Comment:
I think this should actually be outside the `do_query()` method. The behavior of `requestId` is a bit different from `jobId`.